### PR TITLE
Edu benefits controller and model

### LINKED ISSUE #25: Create Education Form Submission Model & Controller

### PR DESCRIPTION
**Pull request by lihanli**
closes #25

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,8 +2,8 @@ language: ruby
 rvm:
   - 2.3.0
 
-services:
-  - postgresql
+addons:
+  postgresql: "9.4"
```
Comment (markolson) [inline .travis.yml]: Let's make sure we double check with devops about the version of pg we'll be using before merging, just in case.

Comment (lihanli) [inline .travis.yml]: i changed the version because we're using a json field in our model, we will have to change it to string if we want to be on an old version

**Comment by markolson** [comment]
LGTM

File: .travis.yml (hunk 1 above)
Comment (markolson) [inline .travis.yml]: Right - I'm just thinking that 9.5 is out, and 9.6 is in beta and may be out soon, and we'll want to verify so our tests don't drift from production where we can avoid that.

File: db/schema.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+ActiveRecord::Schema.define(version: 20160824165957) do
+
+  # These are extensions that must be enabled in order to support this database
+  enable_extension "plpgsql"
+
+  create_table "education_benefits_claims", force: :cascade do |t|
+    t.datetime "submitted_at"
+    t.datetime "processed_at"
+    t.json     "form",         null: false
```
Comment (kreek) [inline db/schema.rb]: What level of PII is in this form? Parts of it may need to be encrypted. We'll eventually have a form service that handles stuff like encryption and field commonality - which this may be the start of - just want to make sure we're not storing SSN, bank account #s in plain text.

File: Gemfile.lock [hunk 3]
```diff
@@ -283,9 +286,11 @@ DEPENDENCIES
   rubocop (~> 0.36.0)
   ruby-saml (~> 1.3.0)
   sdoc (~> 0.4.0)
+  shoulda-matchers
```
Comment (saneshark) [inline Gemfile.lock]: can we not go down this route? shoulda-matchers are bad testing practice imo

we can use acceptance tests and unit tests to test validation, using shoulda-matchers is a bad and fragile way to get the same or similar testing done.

File: spec/models/education_benefits_claim_spec.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+RSpec.describe EducationBenefitsClaim, type: :model do
+  let(:attributes) do
+    {
+      form: { foo: true }
+    }
+  end
+  subject { described_class.new(attributes) }
+
+  describe "validations" do
+    it { should validate_presence_of(:form) }
```
Comment (saneshark) [inline spec/models/education_benefits_claim_spec.rb]: it's better to test that subject.valid? returns false in the absence of :form

File: db/schema.rb (hunk 2 above)
Comment (markolson) [inline db/schema.rb]: This is the form that will have SSN & Routing info. As a dev, I'd rather get this in and have something to develop against while we work on figuring out how to store sensitive info before releasing to production and collecting real data.

Comment (kreek) [inline db/schema.rb]: 👍 that's cool as long as it doesn't make it into the wild

File: spec/models/education_benefits_claim_spec.rb (hunk 4 above)
Comment (lihanli) [inline spec/models/education_benefits_claim_spec.rb]: @saneshark if we do it this way we also have to make sure all of subject's other fields are valid which we shouldnt need to care about to test the form field. from looking at the shoulda code it's setting a field to a value, running validations, then seeing if error messages for the field includes the error it wants to see. that seems like a real unit test and not too fragile, what don't you like about their method?

File: Gemfile.lock (hunk 3 above)
Comment (lihanli) [inline Gemfile.lock]: sorry i didn't see your edited comment, i put a binding;pry on this line https://github.com/thoughtbot/shoulda-matchers/blob/547500730dab994b4e06ce97cc521d54bb53f579/lib/shoulda/matchers/active_model/validator.rb#L92
and it's reaching it for my test. it looks like it is testing behavior and not just verifying the macro

File: spec/models/education_benefits_claim_spec.rb (hunk 4 above)
Comment (saneshark) [inline spec/models/education_benefits_claim_spec.rb]: hmmm, i think they used to just verify the macro. I still don't know if it's necessary to introduce a whole other testing framework on top of rspec but, if it makes testing easier / faster, I'm all for that. 

Why wouldn't you also want to test validations on the other attributes that have validations? Or if form is an association, it could be tested independently and then you just stub the test for form to always return true or false.

Shoulda matchers just add a lot of extra cruft, introduce its own sort of mechanism for testing and once it's introduced to a project, people start using its "convenience methods" without considering the rspec way of doing things using stubs and doubles instead (faster and better practice).

Comment (saneshark) [inline spec/models/education_benefits_claim_spec.rb]: you could initialize subject using factory girl build, then specify you want form: nil, that would have the same effect I would think.

Comment (saneshark) [inline spec/models/education_benefits_claim_spec.rb]: also shoulda matchers have broken quite a few times with new rails versions... particularly with the active record validations. It's just extra cruft, unnecessary... the only thing I've found it particularly useful for is when you want to test dependent destroy -- but i'd sooner write my own rspec matcher for something like that rather than leverage one convenience method from a huge trove of others.

Comment (lihanli) [inline spec/models/education_benefits_claim_spec.rb]: it does seem to add a lot of extra complexity for not much code saved so i dont mind getting rid of it

**Comment by saneshark** [comment]
👍 but I'll defer to @markolson

File: app/controllers/v0/education_benefits_claims_controller.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+module V0
```
Comment (carlmjohnson) [inline app/controllers/v0/education_benefits_claims_controller.rb]: I think it makes more sense to have education/v0 than v0/education, since we'll have multiple APIs in one repo and in the future they may be out of sync with some on v1 permanently and others moving up to v3 or whatever.

Comment (markolson) [inline app/controllers/v0/education_benefits_claims_controller.rb]: This makes sense to me too, but I'd like to hear @saneshark's thoughts

File: db/schema.rb (hunk 2 above)
Comment (shauni) [inline db/schema.rb]: For items that need follow-up before releasing to production, please add TODO or FIXME comments with corresponding issue #. Thanks!

File: app/controllers/v0/education_benefits_claims_controller.rb (hunk 5 above)
Comment (carlmjohnson) [inline app/controllers/v0/education_benefits_claims_controller.rb]: /me can't decide if `EducationBenefits::V0Controller` is a good name or a terrible name.

Comment (shauni) [inline app/controllers/v0/education_benefits_claims_controller.rb]: What about Education::V0::ClaimsController? 

Also, namespacing everything under education has the side benefit of keeping all related code under one directory so we can have
- app/controllers/education
- app/models/education
- etc

It's still under discussion whether common/shared endpoints will live at the toplevel v0 or get moved under common/v0 for consistency. I do think we still want to keep the URLs for common endpoints at the top-level (like api.vets.gov/v0/sessions/new)

Comment (saneshark) [inline app/controllers/v0/education_benefits_claims_controller.rb]: I went down the route of doing v1/rx initially, but I tend to agree that rx/v1 makes a bit more sense... for example if MHV changes stuff up completely on their backend in some new version to support new features, we don't want to bump versions for all other apis, just the ones for prescriptions.

File: .travis.yml (hunk 1 above)
Comment (lihanli) [inline .travis.yml]: devops says we are on 9.4.5 in production

File: app/controllers/v0/education_benefits_claims_controller.rb (hunk 5 above)
Comment (austin-martinez) [inline app/controllers/v0/education_benefits_claims_controller.rb]: Why would you need to bump versions for all the other APIs? You could have /v1/rx and /v2/education living alongside /v0/status

Comment (kreek) [inline app/controllers/v0/education_benefits_claims_controller.rb]: What @austin-martinez said, plus `Vx::MyController` is what I've seen most. You should also have the latest version available without the version # in the URL. This smells a bit like premature optimization, the pattern we followed was to not have a Vx module until it's needed. Once you do need it you move the old code down into a module and the latest remains up top. Routing handles the rest (no pun intended) of the modules that haven't changed.

Comment (shauni) [inline app/controllers/v0/education_benefits_claims_controller.rb]: I am coming around to the version being at the top-level. 

A single client app should be able to use the same version (or default/no version) of every endpoint within the API - so if I'm using `/v0/sessions/new` (a common endpoint) to authenticate for rx, it would be weird to then have to go to `/rx/v3/refill` (or even `/v3/rx/refill`). That said- we shouldn't be maintaining a bunch of different versions - bumping a major version should only happen when we make breaking changes, and that has to happen thoughtfully and rarely. Obviously, breaking changes happen more during active development, so all the APIs should be at `v0` right now. Once we're stable we can bump to v1. When we're adding a new API we can develop it at `v0` and have 'latest' (no version) alias to that. That way we're only ever maintaining two versions - the "latest" and whatever is experimental/volatile/early development.

Comment (carlmjohnson) [inline app/controllers/v0/education_benefits_claims_controller.rb]: > That way we're only ever maintaining two versions - the "latest" and whatever is experimental/volatile/early development

Yes. Version numbering is more important for phone apps because old phone clients never totally go away because some people never upgrade, but in our case, depending on how we do deploys, we shouldn't have users more than 1 version behind the times at any given moment.

Comment (robbiethegeek) [inline app/controllers/v0/education_benefits_claims_controller.rb]: https://dsva.slack.com/archives/vetsgov-engineers/p1472140245000985

Pulling my conversation from slack here

**Comment by plusjeff** [comment]
Let's review this at the Engineering meeting tomorrow.

**Comment by plusjeff** [comment]
Engineering Meeting Notes:

- Robbie: /application/version/action (ex: /rx/v0/status) <--- clear to consumer what version of a specific application, clear contract
- Alastair: /version/application/action <--- less syncing across the entire platform, not rapidly changing versions, less flexibility, begins a culture of integrated profile/product
- John: API Keys. Control access, but also can use it to track who is using which version of the API.
- John: Only need to change the version when existing functionality is modified
- John: Providing one version of functionality is probably easier for us to track and keep our heads around
  - Jeff M: ...and document

**Comment by greggersh** [comment]
I want to add a few notes based on discussion we had after the engineering meeting.

1) In looking at other APIs, when the APIs are for wholly different products, like Google Drive, Google AdWords, Blogger, etc. then the convention is to version each API on its own, which makes sense. I could not find an API that presents as a single API that versions that way.
2) Right now, we're building an API for our front-end apps, so versioning is less important, but still a good idea, to have. Versioning becomes critical, as Robbie said, in delineating a contract between client and server. It will become crucial once those outside the Vets.gov team are using the API. I think it's OK to plan for that now and include versioning.
3) Since no one external is going to be using this API right now, we have some flexibility in versioning. v0 can be in flux until we decide to lock it down and move it to v1, which can have a longer life.

The bigger point:

I think API design is really important, and that right now, we're designing our API to reflect VA's internal architecture, which is the exact thing vets.gov is meant to hide from the outside world.

I think we want to put some thought into what kind of API we're building. To me, we're building an API to a Veteran. To be a bit corny, but I think of the API as a Veteran, and each of the calls as a way of interacting with that Veteran. If I need to send the Veteran a message, I call the send message method. If I need to check her prescriptions, I call the prescriptions method. If I need to file a claim, I post to his claim method.

When you think about it this way, it doesn't make sense to version by product, it makes sense to create a single, cohesive API that defines the interface into a Veteran and all the services associated with them. There are also other implications in terms of how we construct the URLs, and I think we should rethink the "application" part of this as well. We're creating an integrated set of services, and if we expose the internal workings of how we do that and make the API user deal with it, we've failed one of our core missions.

I understand the concern of churn on one "application" or method, so here's what I propose. Since the Vets.gov front end is the only client right now, let's start with v0, which will be explicitly in flux and open to changes. We can leave v0 like this for as long as we need to build out services and let them stabilize, and then rev to v1 when we feel like things are stable and/or we need to start working with outside clients. This gives us the benefit of building in versioning from the start which is a good idea, and also leaves us room for changing stuff without having to maintain a whole bunch of versions across "applications" which, to the end user, make no sense and bear no value.

I also think we should think about how we can make these API methods as simple and clear as possible, so something like:

GET /v0/prescriptions - get a Veterans prescriptions 
POST /v0/claims - create a new claim
GET /v0/claims - get all claims for a Veteran

(Note: in all of these calls, the Veteran is implicitly defined by the auth token/session, so there's no need to include it in the URL. I'd be open to a discussion on whether or not we should do something like 'GET /v0/veteran/claims', but again, I think it's implicit).

In other words, make these calls as RESTful as we can, as if we were building a normal RESTful API for a normal Rails app with normal models. Hide all the complexity from the end user. I'm also fairly certain this approach will provide benefits that we're not even thinking about right now.

**Comment by plusjeff** [comment]
@greggersh : Thanks for the thorough comment! Very much agree with this sentiment: 

> ...we're designing our API to reflect VA's internal architecture, which is the exact thing vets.gov is meant to hide from the outside world.

Let's go with the decision to think of APIs in this way:

<HTTP METHOD> +  /v<version>/<action>

Example: `GET /v0/prescriptions - get a Veterans prescriptions` (from Greg's example)

Additionally, any new/updated API calls should be documented in an `docs/API_<version>.md` file that includes. This file is at the root of `vets-api`:

- Name
- Purpose
- HTTP Method
- URL
- Parameters (required and optional), including validation guidelines
- Example Request
- Example Response

**Comment by plusjeff** [comment]
@lihanli : Can you update this PR per the team decision? From there, you can assign @markolson to review to make sure it's inline with the new standard.
